scvi/inference/jvae_trainer.py

```diff
@@ -0,0 +1,385 @@
+
+    default_metrics_to_monitor = ["elbo"]
+
+    def __init__(
+        self,
+        model: nn.Module,
+        discriminator: nn.Module,
+        gene_dataset_list: List[GeneExpressionDataset],
+        train_size: float = 0.8,
+        use_cuda: bool = True,
+        kappa: float = 1.0,
+        n_epochs_kl_warmup: int = 400,
+        **kwargs
+    ):
+
+        super().__init__(model, gene_dataset_list[0], use_cuda=use_cuda, **kwargs)
+        self.n_epochs_kl_warmup = n_epochs_kl_warmup
+        self.kappa = kappa
+        self.all_dataset = [
+            self.create_posterior(
+                gene_dataset=gd, type_class=partial(JPosterior, mode=i)
+            )
+            for i, gd in enumerate(gene_dataset_list)
+        ]
+        self.n_dataset = len(self.all_dataset)
+        self.all_train, self.all_test = list(
+            zip(
+                *[
+                    self.train_test(
+                        model, gd, train_size, type_class=partial(JPosterior, mode=i)
+                    )
+                    for i, gd in enumerate(gene_dataset_list)
+                ]
+            )
+        )
+        for i, d in enumerate(self.all_train):
+            self.register_posterior("train_%d" % i, d)
+            d.to_monitor = ["elbo"]
+
+        for i, d in enumerate(self.all_test):
+            self.register_posterior("test_%d" % i, d)
+            d.to_monitor = ["elbo"]
+
+        self.discriminator = discriminator
+        if self.use_cuda:
+            self.discriminator.cuda()
+
+        self.kl_weight = None
+        self.compute_metrics_time = None
+        self.n_epochs = None
+
+        self.track_disc = []
+
+    def on_epoch_begin(self):
+        if self.n_epochs_kl_warmup is not None:
+            self.kl_weight = min(1, self.epoch / self.n_epochs_kl_warmup)
+        else:
+            self.kl_weight = 1.0
+
+    @property
+    def posteriors_loop(self):
+        return ["train_%d" % i for i in range(self.n_dataset)]
+
+    def data_loaders_loop(self):
+        posteriors = [self._posteriors[name] for name in self.posteriors_loop]
+        # find the largest dataset to cycle over the others
+        largest = np.argmax(
+            [posterior.gene_dataset.X.shape[0] for posterior in posteriors]
+        )
+
+        data_loaders = [
+            posterior if i == largest else cycle(posterior)
+            for i, posterior in enumerate(posteriors)
+        ]
+
+        return zip(*data_loaders)
+
+    def train(
+        self,
+        n_epochs: int = 20,
+        lr_d: float = 1e-3,
+        lr_g: float = 1e-3,
+        eps: float = 0.01,
+    ):
+        self.compute_metrics_time = 0
+        self.n_epochs = n_epochs
+
+        self.compute_metrics()
+        begin = time.time()
+
+        with torch.set_grad_enabled(True):
+            self.model.train()
+            self.discriminator.train()
+
+            d_params = filter(
+                lambda p: p.requires_grad, self.discriminator.parameters()
```

Comment:
I can but it's done this way in the codebase, and using `filter` make the code more understandable. Why do you want to get rid of this `lambda`? 
It won't cause any problem with `pickle` if that's your concern, because the lambda is not stored in memory.

Reply:
That was my concern!